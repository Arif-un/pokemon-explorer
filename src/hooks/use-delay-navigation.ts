import { useState } from 'react'

import { useRouter } from '@tanstack/react-router'

interface NextLocation {
  to: string
  replace: boolean
}

export function useDelayedNavigation() {
  const router = useRouter()
  const [isNavigating, setIsNavigating] = useState(false)
  const [nextLocation, setNextLocation] = useState<NextLocation | null>(null)

  const navigate = (to: string, { replace = false } = {}) => {
    setIsNavigating(true)
    setNextLocation({ to, replace })
  }

  const finishNavigation = () => {
    if (nextLocation) {
      const { to, replace } = nextLocation
      router.navigate({ to, replace, search: prv => ({ ...prv }) })
      setIsNavigating(false)
      setNextLocation(null)
    }
  }

  return { isNavigating, navigate, finishNavigation }
}
