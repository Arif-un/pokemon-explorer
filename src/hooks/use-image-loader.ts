import { useEffect, useState } from 'react'

interface ImageState {
  src: string
  isLoading: boolean
  error: Event | string | null
}

const imageCache: Record<string, boolean> = {}

export function useImageLoader(src: string, fallbackSrc: string): ImageState {
  const [imageState, setImageState] = useState<ImageState>({
    src: fallbackSrc,
    isLoading: true,
    error: null
  })

  useEffect(() => {
    if (imageCache[src]) {
      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
      return
    }

    const img = new Image()
    img.src = src

    img.onload = () => {
      imageCache[src] = true

      setImageState({
        src: src,
        isLoading: false,
        error: null
      })
    }

    img.onerror = error => {
      setImageState({
        src: fallbackSrc,
        isLoading: false,
        error
      })
    }

    return () => {
      img.onload = null
      img.onerror = null
    }
  }, [src, fallbackSrc])

  return imageState
}
