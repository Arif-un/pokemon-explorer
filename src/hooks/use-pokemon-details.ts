import { useQuery } from '@tanstack/react-query'

import { fetchPokemonDetails } from '@/services/pokemon-api'
import { type PokemonDetailsResponse } from '@/types/Pokemon'

export const usePokemonDetails = <SelectedData = PokemonDetailsResponse>(
  name: string,
  select: (data: PokemonDetailsResponse) => SelectedData
) => {
  const { data, isLoading, isError, error } = useQuery<PokemonDetailsResponse, Error, SelectedData>({
    queryKey: ['pokemon', name],
    queryFn: () => fetchPokemonDetails(name),
    select
  })

  return { data: data as SelectedData, isLoading, isError, error }
}
