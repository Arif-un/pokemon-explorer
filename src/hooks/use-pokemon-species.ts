import { useQuery } from '@tanstack/react-query'

import { fetchPokemonSpecies } from '@/services/pokemon-api'

export const usePokemonSpecies = (name: string) => {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-color', name],
    queryFn: () => fetchPokemonSpecies(name),
    staleTime: 5 * 60 * 1000
  })

  return {
    pokemonColor: data?.color.name,
    isLoading,
    isPokemonSpeciesError: isError,
    pokemonSpeciesError: error
  }
}
