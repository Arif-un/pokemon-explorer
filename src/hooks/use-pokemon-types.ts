import { useQuery } from '@tanstack/react-query'

import { fetchPokemonTypes } from '@/services/pokemon-api-graphql'
import { DEFAULT_QUERY_STALE_TIME } from '@/utils/constants'

export function usePokemonTypes() {
  const { data, isLoading, isError, error } = useQuery({
    queryKey: ['pokemon-types'],
    queryFn: fetchPokemonTypes,
    staleTime: DEFAULT_QUERY_STALE_TIME
  })

  return {
    types: data?.pokemon_v2_type.sort((a, b) => a.name.localeCompare(b.name)),
    isLoading,
    isError,
    error
  }
}
