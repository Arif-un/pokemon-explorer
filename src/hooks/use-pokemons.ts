import { useQuery } from '@tanstack/react-query'

import { fetchPokemonWithFilters } from '@/services/pokemon-api-graphql'
import {
  DEFAULT_POKEMON_FETCH_LIMIT,
  DEFAULT_POKEMON_FETCH_OFFSET,
  DEFAULT_QUERY_STALE_TIME
} from '@/utils/constants'

interface UsePokemonsParams {
  name?: string
  limit?: number
  offset?: number
  types?: string
  sort?: string
}

export const usePokemons = ({
  name,
  limit = DEFAULT_POKEMON_FETCH_LIMIT,
  offset = DEFAULT_POKEMON_FETCH_OFFSET,
  types,
  sort
}: UsePokemonsParams) => {
  const { data, isLoading, isError, error, isFetching } = useQuery({
    queryKey: ['pokemon', name, limit, offset, types, sort],
    queryFn: () => fetchPokemonWithFilters({ name, limit, offset, types, sort }),
    staleTime: DEFAULT_QUERY_STALE_TIME,
    placeholderData: previousData => previousData
  })

  return {
    pokemons: data?.pokemons,
    totalItems: data?.pokemon_aggregate.aggregate.count,
    isLoading,
    isError,
    error,
    isFetching
  }
}
