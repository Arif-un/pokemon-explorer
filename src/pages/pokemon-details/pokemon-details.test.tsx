import { useParams } from '@tanstack/react-router'
import { render, screen, waitFor } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { useDelayedNavigation } from '@/hooks/use-delay-navigation'
import { usePokemonDetails } from '@/hooks/use-pokemon-details'
import { usePokemonSpecies } from '@/hooks/use-pokemon-species'

import PokemonDetails from './pokemon-details'

vi.mock('@/hooks/use-pokemon-details')
vi.mock('@/hooks/use-pokemon-species')
vi.mock('@/hooks/use-delay-navigation')
vi.mock('@tanstack/react-router', () => {
  return {
    useParams: vi.fn()
  }
})

const mockPokemonDetails = {
  id: 25,
  image: 'https://example.com/pikachu.png',
  types: [{ name: 'electric', icon: 'electric-icon.png', color: '#FFD700' }],
  abilities: ['static', 'lightning-rod'],
  baseStats: [
    { name: 'hp', value: 35, max: 255, color: '#FF0000' },
    { name: 'attack', value: 55, max: 255, color: '#F08030' }
  ]
}

describe('PokemonDetails', () => {
  beforeEach(() => {
    vi.mocked(usePokemonDetails).mockReturnValue({
      data: mockPokemonDetails,
      isLoading: false,
      isError: false,
      error: null
    })
    vi.mocked(usePokemonSpecies).mockReturnValue({
      pokemonColor: '#FFD700',
      isLoading: false,
      isPokemonSpeciesError: false,
      pokemonSpeciesError: null
    })
    vi.mocked(useDelayedNavigation).mockReturnValue({
      isNavigating: false,
      navigate: vi.fn(),
      finishNavigation: vi.fn()
    })

    window.scrollTo = vi.fn()

    vi.mocked(useParams).mockReturnValueOnce({ name: 'pikachu' })
  })

  it('renders loading state', () => {
    vi.mocked(usePokemonDetails).mockReturnValueOnce({
      data: undefined,
      isLoading: true,
      isError: false,
      error: null
    })

    render(<PokemonDetails />)

    expect(screen.getByLabelText('Loading skeleton')).toBeInTheDocument()
  })

  it('renders pokemon details correctly', async () => {
    const { asFragment } = render(<PokemonDetails />)

    const { value: hpValue, max: hpMax } = mockPokemonDetails.baseStats[0]
    const { value: attackValue, max: attackMax } = mockPokemonDetails.baseStats[1]

    const hpPercentage = (hpValue / hpMax) * 100
    const attackPercentage = (attackValue / attackMax) * 100

    await waitFor(() => {
      expect(screen.getByText('pikachu')).toBeInTheDocument()
      expect(screen.getByAltText('Pokemon pikachu image')).toHaveAttribute(
        'src',
        'https://example.com/pikachu.png'
      )
      expect(screen.getByText('electric')).toBeInTheDocument()
      expect(screen.getByText('static')).toBeInTheDocument()
      expect(screen.getByText('lightning rod')).toBeInTheDocument()
      expect(screen.getByLabelText(`hp: ${hpPercentage}%`)).toBeInTheDocument()
      expect(screen.getByLabelText(`attack: ${attackPercentage}%`)).toBeInTheDocument()
      expect(asFragment()).toMatchSnapshot()
    })
  })
})
