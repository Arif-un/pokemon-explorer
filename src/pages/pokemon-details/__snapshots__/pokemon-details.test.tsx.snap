// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PokemonDetails > renders pokemon details correctly 1`] = `
<DocumentFragment>
  <div
    class="p-2 mt-52 sm:mt-20 md:mt-20 container mx-auto"
  >
    <div
      class="flex justify-center gap-14 sm:flex-col md:flex-col md:gap-0 sm:gap-0"
    >
      <div>
        <div
          class="sm:fixed md:fixed md:top-[21px] xs:top-[2px] sm:left-4 sm:ml-3 z-50"
          style="opacity: 0; will-change: opacity;"
        >
          <button
            class="px-4 py-2 flex items-center h-11 min-w-11 justify-center outline-none focus:ring-2 ring-yellow-500 bg-slate-50 hover:bg-slate-100 text-slate-900 border-slate-200 dark:bg-slate-900 dark:hover:bg-slate-800 dark:text-slate-100 border dark:border-slate-800 transition rounded-full mb-3"
            type="button"
          >
            <svg
              class="lucide lucide-chevron-left"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m15 18-6-6 6-6"
              />
            </svg>
             Back
          </button>
        </div>
        <div
          class="inner-glow group relative overflow-hidden border border-slate-800 bg-slate-900 dark:bg-opacity-70 p-5 h-[500px] grid place-content-center rounded-3xl"
        >
          <div
            class="h-24 absolute rounded-full m-auto w-full -bottom-28 blur-3xl opacity-40"
            style="background: rgb(255, 215, 0);"
          />
          <div
            class=" "
          >
            <img
              alt="Pokemon image backdrop"
              class="absolute blur-3xl z-0 opacity-50 group-hover:opacity-80 transition-opacity"
              src="https://example.com/pikachu.png"
            />
            <img
              alt="Pokemon pikachu image"
              class="w-96 relative"
              src="https://example.com/pikachu.png"
            />
          </div>
        </div>
      </div>
      <div
        class="mt-9"
      >
        <h2
          class="capitalize text-3xl my-3 font-bold text-slate-950 dark:text-slate-100 mb-5"
        >
          pikachu
        </h2>
        <section>
          <h2
            class="text-slate-600 dark:text-slate-500 mb-1"
            style="opacity: 0; will-change: opacity;"
          >
            Types
          </h2>
          <ul
            class="flex gap-3"
            style="opacity: 0; will-change: opacity;"
          >
            <li
              class="inline-flex justify-center items-center gap-3 p-1 rounded-full"
              style="opacity: 0; will-change: opacity,transform; transform: translateY(15px);"
            >
              <div
                class="rounded-full p-1"
                style="background: rgb(255, 215, 0); box-shadow: 0 4px 20px -2px #FFD700;"
              >
                <img
                  alt="electric type icon"
                  class="size-5"
                  src="electric-icon.png"
                />
              </div>
              <span
                class="capitalize font-semibold mr-3 dark:text-slate-100"
              >
                electric
              </span>
            </li>
          </ul>
        </section>
        <section
          class="my-4"
        >
          <h2
            class="text-slate-600 dark:text-slate-500 mb-1"
            style="opacity: 0; will-change: opacity;"
          >
            Abilities
          </h2>
          <ul
            aria-label="List of Pokemon Abilities"
            class="flex gap-3 w-[370px] sm:w-full flex-wrap"
            style="opacity: 0; will-change: opacity;"
          >
            <li
              class="capitalize flex gap-2 items-center py-1 px-3 text-indigo-700 bg-indigo-200 dark:bg-indigo-900 dark:text-indigo-100 rounded-full font-semibold"
              style="opacity: 0; will-change: opacity,transform; transform: translateY(15px);"
            >
              <svg
                class="size-5"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.78307 2.82598L12 1L20.2169 2.82598C20.6745 2.92766 21 3.33347 21 3.80217V13.7889C21 15.795 19.9974 17.6684 18.3282 18.7812L12 23L5.6718 18.7812C4.00261 17.6684 3 15.795 3 13.7889V3.80217C3 3.33347 3.32553 2.92766 3.78307 2.82598ZM12 13.5L14.9389 15.0451L14.3776 11.7725L16.7553 9.45492L13.4695 8.97746L12 6L10.5305 8.97746L7.24472 9.45492L9.62236 11.7725L9.06107 15.0451L12 13.5Z"
                />
              </svg>
              static
            </li>
            <li
              class="capitalize flex gap-2 items-center py-1 px-3 text-indigo-700 bg-indigo-200 dark:bg-indigo-900 dark:text-indigo-100 rounded-full font-semibold"
              style="opacity: 0; will-change: opacity,transform; transform: translateY(15px);"
            >
              <svg
                class="size-5"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.78307 2.82598L12 1L20.2169 2.82598C20.6745 2.92766 21 3.33347 21 3.80217V13.7889C21 15.795 19.9974 17.6684 18.3282 18.7812L12 23L5.6718 18.7812C4.00261 17.6684 3 15.795 3 13.7889V3.80217C3 3.33347 3.32553 2.92766 3.78307 2.82598ZM12 13.5L14.9389 15.0451L14.3776 11.7725L16.7553 9.45492L13.4695 8.97746L12 6L10.5305 8.97746L7.24472 9.45492L9.62236 11.7725L9.06107 15.0451L12 13.5Z"
                />
              </svg>
              lightning rod
            </li>
          </ul>
        </section>
        <section>
          <h2
            class="text-slate-600 dark:text-slate-500 mb-1"
            style="opacity: 0; will-change: opacity;"
          >
            Stats
          </h2>
          <ul
            aria-label="List of Pokemon Base Stats"
            class="grid grid-cols-3 gap-y-6 gap-x-4 border dark:border-slate-800 bg-slate-100 bg-opacity-70 dark:bg-slate-900 p-4 rounded-xl"
            style="opacity: 0; width: 200px; will-change: opacity;"
          >
            <li
              style="opacity: 0; will-change: opacity,transform; transform: translateY(15px);"
            >
              <div
                class="inline-flex justify-center flex-col items-center min-w-24"
              >
                <figure
                  class="size-14 rounded-full bg-slate-400 bg-opacity-60 dark:bg-slate-900 grid place-content-center relative"
                >
                  <div
                    aria-label="hp: 13.725490196078432%"
                    class="size-14 rounded-full"
                    role="img"
                  />
                  <div
                    class="size-11 rounded-full bg-slate-100 dark:bg-slate-900 absolute inset-0 m-auto grid place-content-center"
                    style="color: rgb(255, 0, 0);"
                  >
                    <svg
                      class="size-6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="filter: drop-shadow(0 2px 7px color-mix(in srgb, #FF0000 70%, transparent));"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M14 20.408c-.492.308-.903.546-1.192.709-.153.086-.308.17-.463.252h-.002a.75.75 0 0 1-.686 0 16.709 16.709 0 0 1-.465-.252 31.147 31.147 0 0 1-4.803-3.34C3.8 15.572 1 12.331 1 8.513 1 5.052 3.829 2.5 6.736 2.5 9.03 2.5 10.881 3.726 12 5.605 13.12 3.726 14.97 2.5 17.264 2.5 20.17 2.5 23 5.052 23 8.514c0 3.818-2.801 7.06-5.389 9.262A31.146 31.146 0 0 1 14 20.408Z"
                      />
                    </svg>
                  </div>
                  <span
                    aria-label="Pokemon stat name"
                    aria-live="polite"
                    class="absolute m-auto right-0 left-0 size-4 -bottom-2 text-xs text-slate-600 dark:text-slate-400"
                  >
                    35
                  </span>
                </figure>
                <figcaption
                  class="text-sm text-center text-slate-600 dark:text-slate-400 capitalize mt-3 font-semibold"
                >
                  hp
                </figcaption>
              </div>
            </li>
            <li
              style="opacity: 0; will-change: opacity,transform; transform: translateY(15px);"
            >
              <div
                class="inline-flex justify-center flex-col items-center min-w-24"
              >
                <figure
                  class="size-14 rounded-full bg-slate-400 bg-opacity-60 dark:bg-slate-900 grid place-content-center relative"
                >
                  <div
                    aria-label="attack: 21.568627450980394%"
                    class="size-14 rounded-full"
                    role="img"
                  />
                  <div
                    class="size-11 rounded-full bg-slate-100 dark:bg-slate-900 absolute inset-0 m-auto grid place-content-center"
                    style="color: rgb(240, 128, 48);"
                  >
                    <svg
                      class="size-6"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      style="filter: drop-shadow(0 2px 7px color-mix(in srgb, #F08030 70%, transparent));"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M7.04813 13.4061L10.5831 16.9421L9.1703 18.3558L10.5849 19.7711L9.17064 21.1853L6.69614 18.71L3.86734 21.5388L2.45312 20.1246L5.28192 17.2958L2.80668 14.8213L4.22089 13.4071L5.63477 14.8202L7.04813 13.4061ZM2.99907 3L6.54506 3.00335L18.3624 14.8207L19.7772 13.4071L21.1915 14.8213L18.7166 17.2962L21.545 20.1246L20.1308 21.5388L17.3024 18.7104L14.8275 21.1853L13.4133 19.7711L14.8269 18.3562L3.00181 6.53118L2.99907 3ZM17.4563 3.0001L20.9991 3.00335L21.001 6.52648L16.9481 10.5781L13.4121 7.0431L17.4563 3.0001Z"
                      />
                    </svg>
                  </div>
                  <span
                    aria-label="Pokemon stat name"
                    aria-live="polite"
                    class="absolute m-auto right-0 left-0 size-4 -bottom-2 text-xs text-slate-600 dark:text-slate-400"
                  >
                    55
                  </span>
                </figure>
                <figcaption
                  class="text-sm text-center text-slate-600 dark:text-slate-400 capitalize mt-3 font-semibold"
                >
                  attack
                </figcaption>
              </div>
            </li>
          </ul>
        </section>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
